driver-sync/src/test/functional/com/mongodb/internal/connection/OidcAuthenticationProseTests.java

```diff
@@ -256,118 +253,210 @@ public void test3p1AuthFailsWithCachedToken() throws ExecutionException, Interru
-        MongoClientSettings clientSettings = createSettings(getOidcUri(), onRequest);
+        TestCallback callback = createCallback();
+        MongoClientSettings clientSettings = createSettings(callback);
         try (MongoClient mongoClient = createMongoClient(clientSettings)) {
             failCommand(391, 1, "find");
             // #. Perform a find operation that succeeds.
             performFind(mongoClient);
         }
-        assertEquals(2, onRequest.invocations.get());
+        assertEquals(2, callback.invocations.get());
+    }
+
+    @Test
+    public void test4p2ReadCommandsFailIfReauthenticationFails() {
+        // Create a `MongoClient` whose OIDC callback returns one good token
+        // and then bad tokens after the first call.
+        TestCallback wrappedCallback = createCallback();
+        OidcCallback callback = (context) -> {
+            OidcCallbackResult result1 = wrappedCallback.callback(context);
+            return new OidcCallbackResult(wrappedCallback.getInvocations() > 1 ? "bad" : result1.getAccessToken());
+        };
+        MongoClientSettings clientSettings = createSettings(callback);
+        try (MongoClient mongoClient = createMongoClient(clientSettings)) {
+            performFind(mongoClient);
+            failCommand(391, 1, "find");
+            assertCause(MongoCommandException.class,
+                    "Command failed with error 18",
+                    () -> performFind(mongoClient));
+        }
+        assertEquals(2, wrappedCallback.invocations.get());
+    }
+
+    @Test
+    public void test4p3WriteCommandsFailIfReauthenticationFails() {
+        // Create a `MongoClient` whose OIDC callback returns one good token
+        // and then bad tokens after the first call.
+        TestCallback wrappedCallback = createCallback();
+        OidcCallback callback = (context) -> {
+            OidcCallbackResult result1 = wrappedCallback.callback(context);
+            return new OidcCallbackResult(
+                    wrappedCallback.getInvocations() > 1 ? "bad" : result1.getAccessToken());
+        };
+        MongoClientSettings clientSettings = createSettings(callback);
+        try (MongoClient mongoClient = createMongoClient(clientSettings)) {
+            performInsert(mongoClient);
+            failCommand(391, 1, "insert");
+            assertCause(MongoCommandException.class,
+                    "Command failed with error 18",
+                    () -> performInsert(mongoClient));
+        }
+        assertEquals(2, wrappedCallback.invocations.get());
+    }
+
+    private static void performInsert(final MongoClient mongoClient) {
+        mongoClient
+                .getDatabase("test")
+                .getCollection("test")
+                .insertOne(Document.parse("{ x: 1 }"));
+    }
+
+    @Test
+    public void test5p1Azure() {
+        assumeTrue(getOidcEnv().equals("azure"));
+        String oidcUri = getOidcUri();
+        assertFalse(oidcUri.contains("@"));
```

Comment:
It is not obvious that this check is to make sure `getOidcUri()` does not have a user name specified. Let's rather do this check using `clientSettings`, where it will be more clear and less error-prone.